### ISSUE TITLE
Strange distortion when automating CLAP plug-in parameters in DAW?

### ISSUE THREAD

**Issue by lewloiwc** [free-audio/clap#324]
https://github.com/baconpaul/clap-c99-distortion
https://nakst.gitlab.io/tutorial/clap-part-2.html
I tried both of these in Bitwig and they behaved as I expected with direct mouse modification of parameters and modulation by modulation, but with automation I got this strange distortion.
![image](https://github.com/free-audio/clap/assets/87447432/5e5ffeb8-6617-4555-9588-ec9fd2d3f586)

When I test the CLAP plugin in REAPER, it looks ok at first glance, but the automation in the moment of playback part does not seem to be working.
![image](https://github.com/free-audio/clap/assets/87447432/8da4ad4d-3457-4204-b65c-47c3c9f4b358)

Is the code for these two CLAP plugins written correctly?
Or is the problem with your DAW such as Bitwig or REAPER?

What I'm imagining:
-The code in the two Clap plugins is fine
-Bitwig's CLAP plugin automation is buggy
-REAPER's CLAP plugin has a bug where the automation of the part at the moment of playback does not work.

Translated with www.DeepL.com/Translator (free version)

**Comment by lewloiwc** [comment]
I feel certain that there is at least a bug in Bitwig.

Purple is the value of the parameter and green is the number of times CLAP_EVENT_PARAM_VALUE was called.
The manual operation was successfully called only once for BUFFER SIZE.
![image](https://github.com/free-audio/clap/assets/87447432/1d4fd3a9-89b2-4d37-950f-ab5cb9b5b3b8)

Modulation was also successfully called every 128 samples in buffer size.
![image](https://github.com/free-audio/clap/assets/87447432/e3ce4c7e-1bfb-42a1-8664-6f9cc064e87c)

However, automation calls every 128 samples in the first buffer, then calls only once in buffer size, and after a while it calls every 128 samples again.
The parameter values are also unstable.
![image](https://github.com/free-audio/clap/assets/87447432/8009e05f-5c77-4f15-9d50-c603ed26a5d5)

What do you think?

Translated with www.DeepL.com/Translator (free version)

**Comment by lewloiwc** [comment]
https://github.com/free-audio/interop-tracker/issues
I just found this just now.
I should probably post these issues over there, right?
Can I move the issues?
Or should I close the issue and repost it over there?

Translated with www.DeepL.com/Translator (free version)

**Comment by baconpaul** [comment]
it probably does belong in the inter-op tracker yes.
which version of bitwig and which OS would also help I am sure
Does surge exhibit the same behavior? That's different code than the c99 distortion. But I looked at the c99 distortion and i looks alright (albeit a bit convoluted) in how it processes events. I would not rule out a bug there though.

**Comment by lewloiwc** [comment]
Windows 10 Home
Bitwig Studio 4.4.10
Sample Rate: 96000 Hz / Buffer Size: 1024 Samples
Surge XT 1.2.3.bb8dd2b

Also, when I was trying to check the Surge XT version, I noticed that you set the buffer size to 1024 samples, but Surge XT recognizes it as 32 samples, isn't that strange?
![image](https://github.com/free-audio/clap/assets/87447432/abc7d300-6157-46c4-bce3-2d1522b0718a)
![image](https://github.com/free-audio/clap/assets/87447432/edde9f49-3d5a-4c7d-a05a-f721da2e4b51)

I checked Surge XT and found that this also had the same distortion when using automation.
![image](https://github.com/free-audio/clap/assets/87447432/9a7aed01-b59e-486e-adb9-676107ea43b5)

Modulation is still fine, but I am concerned that Surge XT is doing linear interpolation with 32 samples, probably because it recognizes the buffer size as 32 samples.
![image](https://github.com/free-audio/clap/assets/87447432/19eb0d66-7b81-403a-b81f-ae8cc8459b29)

**Comment by baconpaul** [comment]
That’s correct. Production surge has an internal block of 32 and uses the sample time of events to interneave with profess calls appropriately no matter the parent block size 

how are you making these charts?

**Comment by lewloiwc** [comment]
I didn't know Surge was designed that way! Thanks for letting me know!

The chart is made by setting the Surge XT oscillator to a high frequency sine wave, automating A Osc 1 Level, and viewing the output audio on a ReaJS oscilloscope.

I also asked Bitwig Support and they said "We are not aware of a problem with commercially released CLAP plug-ins,
I checked u-he MFM2, a commercially released CLAP plug-ins, and found the same distortion.

First, output an audio file with this setup.
Windows 10 Home
Bitwig Studio 4.4.10
Sample Rate: 96000 Hz / Buffer Size: 1024 Samples
MFM2 2.5 (revision 13385) (CLAP)
![image](https://github.com/free-audio/clap/assets/87447432/727af689-92a0-4ffd-991e-444dd37d1147)
bwproject→https://drive.google.com/file/d/1-ZLAaBPHRIkGOXoF44fjfQpnjtPzk0qU/view?usp=sharing

Next, check the output audio file with an oscilloscope. The shape now looks smooth and problem-free, but that is because MFM2 has smoothed the values with a 1-pole low-pass filter.
![image](https://github.com/free-audio/clap/assets/87447432/c6da1527-6065-405d-bdde-9cc4bae0a2d4)

This audio file is then used to calculate the curvature using the following calculation.
B = A;
A = input;
output = A - B;
![image](https://github.com/free-audio/clap/assets/87447432/4f6a9dd8-8fee-4e2f-8601-43abd903b58a)

If you zoom in on this, it will look like this
![image](https://github.com/free-audio/clap/assets/87447432/2b99c321-e82b-402e-8213-e155aec0cf56)

Isn't this exactly the same distortion as this one? The shape of the distortion and the timing of the distortion are also the same?
![image](https://github-production-user-asset-6210df.s3.amazonaws.com/87447432/238121975-5e5ffeb8-6617-4555-9588-ec9fd2d3f586.png)
